### PR TITLE
WT-9252 instantiated fast-truncate pages in read-only trees fail eviction tests

### PR DESCRIPTION
**Pull request by keithbostic**

### CODE REVIEW

**Comment by sauclovian-wt** [comment]
LGTM

**Review by sauclovian-wt — APPROVED** [review]
LGTM

**Comment by keithbostic** [comment]
@kommiharibabu, @sauclovian-wt, @sulabhM:

This change is small, but it's got some scary bits.  There's a bit of random noise, but there are 3 significant changes:

David, the change in `bt_read.c:__page_read()` is what fixes the checkpoint cursor read test. If a page gets evicted, it has a state of WT_REF_DISK and a fast-truncate structure, and when we read the page back in we don't instantiate it as an in-memory page because the WT_REF.state doesn't match the test. I think we can simply test the WT_REF.ft_info.del field, as I believe that can only be set if the WT_REF.state is WT_REF_DELETED or WT_REF_DISK.

David, the change in `btree_inline.h:__wt_page_can_evict()` is a rewrite from what we originally discussed: rather than testing the WT_BTREE's readonly status, I'm checking if the page is "clean". I don't think that can happen with a fast-truncate structure (unless it's a checkpoint read), but it shouldn't hurt, either.

Sulabh, the change in `btree_inline.h: __wt_btcur_skip_page()` allows us to skip fast-truncate pages when reading from a checkpoint cursor. The change is to check the fast-truncate information instead of the written page's cell timestamps: that should allow a transaction to skip its own fast-truncate operations, as well. I don't think this change is necessary, but it should speed up cursor traversals in some workloads. Note I removed the check for WT_REF_DELETED in the more general test of the page's timestamps. I'm also very concerned about the visibility check I added in that code, please make sure I didn't get that wrong.

Thanks!

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Comment by sulabhM** [comment]
> Note I removed the check for WT_REF_DELETED in the more general test of the page's timestamps.

Is that because a truncated page if initialized will continue to have `ft_info>del` populated, as mentioned in the comment about `ft_info`:

`If the locked WT_REF's previous state was WT_REF_DELETED, WT_REF.del is valid, if the WT_REF's previous state was an in-memory state, then WT_REF.update is valid.`

**Comment by keithbostic** [comment]
> Is that because a truncated page if initialized will continue to have `ft_info>del` populated

The instantiated page will be in state WT_REF_MEM, I believe, not WT_REF_DELETED. What I was thinking there is that if the state is WT_REF_DELETED, there must be a ft_info.del value, in all cases, so there's no point in checking it again.

**Comment by sauclovian-wt** [comment]
LGTM

File: src/include/btree_inline.h [hunk 1]
```diff
@@ -2073,45 +2077,74 @@ __wt_btcur_skip_page(
...
+     *
+     * #2 and #3 may have ft_info.del set, but it may be NULL, so we check.
+     */
+    page_del = NULL;
+    if (previous_state == WT_REF_DELETED)
+        page_del = ref->ft_info.del;
+    if (F_ISSET(btree, WT_BTREE_READONLY) &&
+      (previous_state == WT_REF_DISK || previous_state == WT_REF_MEM))
+        page_del = ref->ft_info.del;
+    if (page_del != NULL && __wt_txn_visible(session, page_del->txnid, page_del->timestamp)) {
```
Comment (agorrod) [inline src/include/btree_inline.h:2129]: It seems odd to me that we would skip pages that are visible to the current session. Is this check the right way around?

Comment (kommiharibabu) [reply src/include/btree_inline.h:2129]: We are skipping the page where the fast truncate operation is visible to us, so there are no updates on the page that are visible to us.

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Review by sulabhM — COMMENTED** [review]
LGTM

**Pull request merged**